src/inference/ns.jl

```diff
@@ -0,0 +1,354 @@
+        vectorize(dist, mean(dist))
+    end
+
+    NSModel(model, spl, μ)    ## or NestedModel ?? check
+end
+
+function AbstractMCMC.sample_end!(
+    rng::AbstractRNG,    ## check ??
+    model::Model,     ## or `AbstractModel`
+    spl::Sampler{<:NS},      ## or `Nested`
+    N::Integer,
+    transitions;
+    debug::Bool = false,
+    kwargs...
+)
+    # Pop remaining points in ellipsoid
+    N = length(transitions)
+    spl.log_vol = -N / spl.nactive - log(spl.nactive)
+    @inbounds for i in eachindex(spl.active_logl)
+        # get new point
+        draw = spl.active_points[:, i]
+        logL = spl.active_logl[i]
+        log_wt = spl.log_vol + logL
+
+        # update sampler
+        logz = logaddexp(spl.logz, log_wt)
+        spl.h = (exp(log_wt - logz) * logL +
+               exp(spl.logz - logz) * (spl.h + spl.logz) - logz)
+        spl.logz = logz
+
+        prev = NestedTransition(draw, logL, log_wt)
+        push!(transitions, prev)
+    end
+
+    # h should always be non-negative. Numerical error can arise from pathological corner cases
+    if spl.h < 0
+        spl.h ≉ 0 && @warn "Negative h encountered h=$(spl.h). This is likely a bug"
+        spl.h = zero(spl.h)
+    end
+
+    return nothing
+end
+
+function AbstractMCMC.bundle_samples(    ## 2 `AbstractMCMC.bundle_samples` functions
+    ::AbstractRNG,  
+    model::Model,    ## or `AbstractModel`
+    spl::Sampler{<:NS},    ## or `Nested`
+    ::Integer,
+    transitions,
+    Chains;
+    param_names = missing,
+    check_wsum = true,
+    kwargs...
+)
+
+    vals = copy(mapreduce(t->vcat(t.draw, t.log_wt), hcat, transitions)')
+    # update weights based on evidence
+    @. vals[:, end, 1] = exp(vals[:, end, 1] - spl.logz)
+    wsum = sum(vals[:, end, 1])
+    @. vals[:, end, 1] /= wsum
+
+    if check_wsum
+        err = !iszero(spl.h) ? 3sqrt(spl.h / spl.nactive) : 1e-3
+        isapprox(wsum, 1, atol = err) || @warn "Weights sum to $wsum instead of 1; possible bug"
+    end
+
+    # Parameter names
+    if param_names === missing
+        param_names = ["Parameter $i" for i in 1:length(vals[1, :]) - 1]
+    end
+    push!(param_names, "weights")
+
+    return Chains(vals, param_names, Dict(:internals => ["weights"]), evidence = spl.logz)
+end
+
+function AbstractMCMC.bundle_samples(    ## 2 `AbstractMCMC.bundle_samples` functions
+    rng::AbstractRNG,   
+    model::Model,    ## or `AbstractModel`
+    spl::Sampler{<:NS},     ## or `Nested`
+    ::Integer,
+    transitions,
+    A::Type{<:AbstractArray};
+    check_wsum = true,
+    kwargs...
+)
+
+    vals = convert(A, mapreduce(t->t.draw, hcat, transitions)')
+
+    if check_wsum
+        # get weights
+        wsum = mapreduce(t->exp(t.log_wt - spl.logz), +, transitions)
+
+        # check with h
+        err = spl.h ≠ 0 ? 3sqrt(spl.h / spl.nactive) : 1e-3
+        isapprox(wsum, 1, atol = err) || @warn "Weights sum to $wsum instead of 1; possible bug"
+    end
+
+    return vals
+end
```

Comment:
None of this is necessary. If you are correctly returning a `Transition`, the rest of Turing will handle this for you.